File: languages/go/client/availability.go
package client

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/availability"
	"github.com/bloock/bloock-sdk-go/v2/entity/record"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

// AvailabilityClient represents a client for interacting with the [Bloock Availability service].
//
// [Bloock Availability service]: https://bloock.com
type AvailabilityClient struct {
	bridgeClient bridge.BloockBridge
	configData   *proto.ConfigData
}

// NewAvailabilityClient creates a new instance of the AvailabilityClient with default configuration.
func NewAvailabilityClient() AvailabilityClient {
	return AvailabilityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   config.NewConfigDataDefault(),
	}
}

// NewAvailabilityClientWithConfig creates a new instance of the AvailabilityClient with the provided configuration.
func NewAvailabilityClientWithConfig(configData *proto.ConfigData) AvailabilityClient {
	return AvailabilityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   configData,
	}
}

// Publish publishes a Bloock record to the Availability service using the specified publisher.
func (c *AvailabilityClient) Publish(r record.Record, publisher availability.Publisher) (string, error) {
	res, err := c.bridgeClient.Availability().Publish(context.Background(), &proto.PublishRequest{
		ConfigData: c.configData,
		Record:     r.ToProto(),
		Publisher:  publisher.ToProto(),
	})

	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return res.Id, nil
}

// Retrieve retrieves a Bloock record from the Availability service using the specified loader.
func (c *AvailabilityClient) Retrieve(loader availability.Loader) (record.Record, error) {
	res, err := c.bridgeClient.Availability().Retrieve(context.Background(), &proto.RetrieveRequest{
		ConfigData: c.configData,
		Loader:     loader.ToProto(),
	})

	if err != nil {
		return record.Record{}, err
	}

	if res.Error != nil {
		return record.Record{}, errors.New(res.Error.Message)
	}

	return record.NewRecordFromProto(res.Record, c.configData), nil
}

File: languages/js/src/client/availability.ts
import { BloockBridge } from "../bridge/bridge";
import {
  PublishRequest,
  RetrieveRequest
} from "../bridge/proto/bloock_availability";
import { ConfigData } from "../bridge/proto/bloock_config";
import { NewConfigData } from "../config/config";
import { Loader, Publisher, PublishResponse } from "../entity/availability";
import { Record } from "../entity/record";

/**
 * Represents a client for interacting with the [Bloock Availability service](https://dashboard.bloock.com/login).
 */
export class AvailabilityClient {
  private bridge: BloockBridge;
  private configData: ConfigData;

  /**
   * Creates a new instance of the AvailabilityClient with default configuration.
   * @param configData
   */
  constructor(configData?: ConfigData) {
    this.bridge = new BloockBridge();
    this.configData = NewConfigData(configData);
  }

  /**
   * Publishes a Bloock record to the Availability service using the specified publisher.
   * @param record
   * @param publisher
   * @returns
   */
  async publish(record: Record, publisher: Publisher): Promise<PublishResponse> {
    const request = PublishRequest.fromPartial({
      configData: this.configData,
      publisher: publisher.toProto(),
      record: record.toProto()
    });
    return this.bridge
      .getAvailability()
      .Publish(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return PublishResponse.fromProto(res);
      });
  }

  /**
   * Gets a Bloock record from the Availability service using the specified loader.
   * @param loader
   * @returns
   */
  async retrieve(loader: Loader): Promise<Record> {
    const request = RetrieveRequest.fromPartial({
      configData: this.configData,
      loader: loader.toProto()
    });
    return this.bridge
      .getAvailability()
      .Retrieve(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Record.fromProto(res.record!, this.configData);
      });
  }
}
